chapters/os_tooling.md

```diff
@@ -0,0 +1,85 @@
+# Offline Compilation of OpenCL Kernel Sources
+
+Aside from online compilation during the application execution, OpenCL kernel sources can be compiled offline into binaries that can be loaded into the drivers using special API (e.g. `clCreateProgramWithBinary` or `clCreateProgramWithIL`).
+
+This section described available open source tools for offline compilation of OpenCL kernels.
+
+## Open Source Tools
+
+* [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages including OpenCL C and C++ for OpenCL that can compile into an executable binary (e.g. AMDGPU), or a portable binary (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).
+* [libclc](https://libclc.llvm.org/) is a generic and portable implementation of OpenCL builtin function libraries for OpenCL 1.1 - and some functions from later versions of OpenCL can be found there too.
+* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) is a library and __llvm-spirv__ tool for translating between LLVM IR and SPIR-V. 
+* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
+* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.
+
+### Compiling Kernels to SPIR-V
+
+Available open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.
+
+<p align="center">
+<br>
+<img src="../images/opencl_to_spirv_tooling.jpg">
```

Comment:
> A nit about the image: I would only mention .clcpp, not .cl/.clcpp, for C++ for OpenCL, to encourage its adoption. This is of course conditional on the corresponding PR in LLVM. Unless we consider this might be too new to be available?

I think you make a good point and the change was already merged. In the text I do have another option using _.cl_ extension so I think it is fine to keep _.clcpp_ only in the diagram. This should be the flow we advertise anyway.

> A possible point of confusion: the two flavors of SPIR-V are not mentioned in this document, nor in the figure. It might be misleading in that someone could think a Vulkan SPIR-V module can be linked with an OpenCL SPIR-V module. I'm not sure how important that is, in the context of this document.

Yeah, I agree we should avoid such confusion. So there are two changes I could think of:
1. We can add a footnote in the diagram for `spirv-link` box explaining that it is not possible to link modules from two dialects
2. We can just add a note in the text of `spirv-link` example.

Any preferences or other suggestions?

Reply:
> A possible point of confusion: the two flavors of SPIR-V are not mentioned in this document...

Having read the markdown version before discussions, I would like to second that. Not just the linking part, but the image may be interpreted that Vulkan compute shaders could be authored using C++ for OpenCL and use the OpenCL flavor of the toolchain to compile them to "just" SPIR-V (as opposed to compute SPIR-V).

Reply:
> Vulkan compute shaders could be authored using C++ for OpenCL

Just to make sure I understood you. Do you mean that the figure might be misleading someone to think that Vulkan shaders can be written in C++ for OpenCL?

>  and use the OpenCL flavor of the toolchain to compile them to "just" SPIR-V (as opposed to compute SPIR-V).

Do you suggest that we should indicate somehow on the diagram that clspv path uses Vulkan flavour SPIR-V and the other path uses OpenCL flavour SPIR-V? Or do you think it is sufficient to explain it in the text?

Reply:
> Any preferences or other suggestions?

To be honest, I'm not sure what the right solution is -- and I'm personally happy if this is dealt with at a later time. It might be better to refer explicitly to each SPIR-V flavor in the figure to avoid any confusion. But at the same time I was surprised that the documentation of the SPIR-V Tools doesn't mention the incompatibility of the two flavors so maybe a more global clarification would be useful.

Reply:
> It might be better to refer explicitly to each SPIR-V flavor in the figure to avoid any confusion.

This can be done in a top part but then there is a common part and I don't have a good way to achieve this apart from just drawing the teo flows completely separtely. Does it make sense?

Otherwise, I will explain it in the text for now. Actually, I think it should be explained in the text anyway. Then we can see if the diagram needs to be updated or we can postpone it and solve in follow up PRs.

Reply:
I have added a paragraph below the diagram. Do you think it is good enough for now?

Reply:
> I have added a paragraph below the diagram. Do you think it is good enough for now?

Looks good to me.